### ISSUE TITLE
do not handle(ignore) zuluCrypt's encryption mappers and mount points

### ISSUE THREAD

**Issue by mhogomchungu** [storaged-project/udisks#480]
zuluCrypt[1] creates device mappers with prefix "/dev/mapper/zuluCrypt-XXXX" and mount points with prefixes of "/run/media/private/XXX" and "/run/media/public/YYY" and i would like udisks to ignore devices and mount paths that have these properties because udisks' handling of these volumes creates problems like this[2] and this[3].

[1] https://mhogomchungu.github.io/zuluCrypt/
[2] https://github.com/mhogomchungu/zuluCrypt/issues/84
[3] http://www.pclinuxos.com/forum/index.php/topic,134802.msg1148264.html#msg1148264

**Comment by pothos** [comment]
Hello,
there is a similar request regarding Snap. The idea was to add a new mount option that can be seen in UDisks via a mtab property. Then this would be queried in GNOME Disks or the respective KDE application to not show these objects. Andrea was working on this:
https://bugzilla.gnome.org/show_bug.cgi?id=790279
https://github.com/storaged-project/udisks/pull/460
Would this also solve your issue or is it different?

**Comment by vojtechtrefny** [comment]
Udisks doesn't automatically mount devices, something has to tell us to mount a new device -- e.g. gvfs in Gnome (I don't know what tool does this in KDE). So I guess this should be fixed/changed in the tool that tells us to mount the device.
We can do some things to help (like adding the property with mount options as @pothos mentioned that can tell gvfs/kde to not mount it) but if someone asks us to mount a device we can't refuse to do that just because it is already mounted.

**Comment by mhogomchungu** [comment]
I think what happens is:
1. zuluCrypt creates an encryption mapper.
2. udisks notices the created encryption mapper and announces its presence.
3. Other tools like device notifier in the reported case notices the announcement and asks udisks to mount the encryption mapper.
4. udisks comply creating two mount points causes zuluCrypt to get confused.

Here[1] is another case where a problem is reported on zuluCrypt because dolphin(KDE's default file manager) shows zuluCrypt encryption mapper on its side panel. There is a whole lot of confusion in that thread but the source of it is that a user is trying to manage a zuluCrypt encryption mapper using udisks through dolphin. This shows that the trick of not showing mount points will not be effective.

I think a workable solution is for udisks to refuse to mount a device that is ready mounted. Can you think of a reason of why a person would want to mount the same device more than once? Best way to do this as far as i know is to create bind mounts. Of course this will create a race condition i think zuluCrypt will win most of the time because it will mount the volume immediately after the mapper is created while udisks will first do the dance of announcing the presence of the mapper and wait to be asked to mount it.

My initial though was to make udisks ignore zuluCrypt devices through udev rule but the above suggestion seems like a better alternative.

[1] https://classicforum.manjaro.org/index.php?topic=32344.msg264764#msg264764

**Comment by vpodzime** [comment]
> I think a workable solution is for udisks to refuse to mount a device that is ready mounted.

I don't think this policy belongs to *udisks*. Mounting a device multiple times is a supported thing and whatever the potential use cases for this are, it shouldn't be *udisks* deciding that suddenly this is not supported. But maybe a new extra option for the ``Mount()`` method could be added for this?

**Comment by vojtechtrefny** [comment]
> Can you think of a reason of why a person would want to mount the same device more than once?

Btrfs with sobvolumes needs to be mounted multiple times (for each subvolume with a different `subvol` option). You probably won't do that with udisks, but it is a valid use case.

KDE should ask if the device is mounted and not try to mount it again -- but even with this check there is a huge possibility of race conditions.

I think that telling udisks to ignore the devices using udev is the best option here. You are basically replacing our functionality for some subset of devices created (and managed) by zuluCrypt so it makes sense to tell us to ignore them.

**Comment by vojtechtrefny** [comment]
I've just checked our code and we actually don't allow mounting devices that are already mounted -- https://github.com/storaged-project/udisks/blob/master/src/udiskslinuxfilesystem.c#L1272 -- I had no idea we do this.

**Comment by mhogomchungu** [comment]
I have search udisks source code and failed to find the implementation of "udisks_filesystem_get_mount_points" function, could someone point to where it is.

**Comment by vojtechtrefny** [comment]
This function is generated by gdbus-codegen. It just reads the `Mountpoints` property of the `Filesystem` interface on dbus. The property is updated when the `/proc/self/mountinfo` file changes.
So either udisks is faster than zuluCrypt when mounting the device, or it takes too long to get the information about the mountpoint change and the mount function doesn't fail because the `Mountpoints` property is not updated yet.
Even if we change this, there is still a lot of space for race conditions. I still think the best solution would be to ignore the devices "managed" by zuluCrypt using udev rules. (And I also think this rules should be part of zuluCrypt -- it would be easier in case you change something in the future and you have more information about the devices than just a name -- I don't like the idea of adding an "ignore" rule to UDisks to ignore everything based only on a name prefix.)

**Comment by arschficken** [comment]
I should document for completeness that this problem also arises when _another user_ mounts the filesystem. Bug #507 documents that error mode.
